core/src/socket.rs

```diff
@@ -0,0 +1,239 @@
+use crate::{
+    avm2::{Activation, Avm2, EventObject, Object, TObject},
+    backend::navigator::NavigatorBackend,
+    context::UpdateContext,
+};
+use gc_arena::Collect;
+use generational_arena::{Arena, Index};
+use std::collections::VecDeque;
+
+/// Socket backend implementation
+///
+/// When `Socket.connect()` is called, the backend may provide a
+/// [SocketConnection] instance, that will be polled every tick.
+/// Methods must not block.
+///
+/// For performances reasons, only 1 message will be polled per tick per connection
+/// and at most 1 message will be sent per tick per connection.
+///
+/// When the Socket gets closed, the underlying [SocketConnection]
+/// is dropped, so simply implement [Drop] to handle cleanup stuff.
+///
+/// See [SocketConnection::is_connected] for details about how
+/// Ruffle infer connection state, [SocketConnection::send] for details
+/// about forwarding messages, [SocketConnection::poll] for details
+/// about polling incoming messages.
+pub trait SocketConnection {
+    /// Return socket connection state
+    ///
+    /// Possibles values:
+    ///  * `None`: connecting ([SocketConnection::send] and
+    ///            [SocketConnection::poll] not yet called)
+    ///  * `Some(true)`: connected ([SocketConnection::send] and
+    ///                 [SocketConnection::poll] got called)
+    ///  * `Some(false)`: disconnected, or connection refused socket wasn't connected
+    ///                   (called only after [SocketConnection::poll] return `None`;
+    ///                    [SocketConnection::send] and [SocketConnection::poll]
+    ///                    won't be called after `Some(false)` is returned)
+    ///
+    /// Returning `None` after `Some(true)` was returned
+    /// will be considered as `Some(false)` (i.e. connection closed).
+    fn is_connected(&self) -> Option<bool>;
+
+    /// Send a message to the remote side
+    ///
+    /// `buf` contains all bytes for the message to send, it's up to the backend to
+    /// decide how to encode and transport them.
+    ///
+    /// Only called if there is at least 1 pending message to send.
+    fn send(&mut self, buf: Vec<u8>);
+
+    /// Poll the next available message, if any
+    ///
+    /// Called every ticks, so can be used to try to
+    /// flush message not sent yet for instance.
+    fn poll(&mut self) -> Option<Vec<u8>>;
+}
+
+pub type SocketHandle = Index;
+
+#[derive(Collect)]
+#[collect(no_drop)]
+struct Socket<'gc> {
+    target: Object<'gc>,
```

Comment:
Wouldn't it be better to have this as an `SocketObject`, since there's no way it could be any other kind of object?

Reply:
Good idea. I did not think of that.